%% LyX 1.3 created this file.  For more info, see http://www.lyx.org/.
%% Do not edit unless you really know what you are doing.
\documentclass[english, 12pt]{article}
\usepackage{times}
%\usepackage{algorithm2e}
\usepackage{url}
\usepackage{bbm}
\usepackage[T1]{fontenc}
\usepackage[latin1]{inputenc}
\usepackage{geometry}
\geometry{verbose,letterpaper,tmargin=2cm,bmargin=2cm,lmargin=1.5cm,rmargin=1.5cm}
\usepackage{rotating}
\usepackage{color}
\usepackage{graphicx}
\usepackage{amsmath, amsthm, amssymb}
\usepackage{setspace}
\usepackage{lineno}
\usepackage{hyperref}
\usepackage{bbm}
\usepackage{makecell}
\usepackage{placeins}
\usepackage{subcaption}

%\renewcommand{\arraystretch}{1.8}

%\linenumbers
%\doublespacing
\onehalfspacing
\usepackage[numbers,super]{natbib}

%Pour les rajouts
\usepackage{color}
\definecolor{trustcolor}{rgb}{0,0,1}

\usepackage{dsfont}
\usepackage[warn]{textcomp}
\usepackage{adjustbox}
\usepackage{multirow}
\usepackage{subcaption}
\usepackage{graphicx}
\graphicspath{{../figures/}}
\DeclareMathOperator*{\argmin}{\arg\!\min}

\let\tabbeg\tabular
\let\tabend\endtabular
\renewenvironment{tabular}{\begin{adjustbox}{max width=0.95\textwidth}\tabbeg}{\tabend\end{adjustbox}}

\makeatletter

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% LyX specific LaTeX commands.
%% Bold symbol macro for standard LaTeX users
%\newcommand{\boldsymbol}[1]{\mbox{\boldmath $#1$}}

%% Because html converters don't know tabularnewline
\providecommand{\tabularnewline}{\\}
\renewcommand*{\arraystretch}{1.2}

\usepackage{babel}
\makeatother

\newcommand{\NEFF}{Option \texttt{Neff} controls whether a logistic regression is used for the GWAS with the effective sample in LDpred2, or a linear regression and then the total sample size.}

\begin{document}

\renewcommand{\thefigure}{S\arabic{figure}}
\setcounter{figure}{0}
\renewcommand{\thetable}{S\arabic{table}}
\setcounter{table}{0}
\renewcommand{\theequation}{S\arabic{equation}}
\setcounter{equation}{0}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[p]
\centerline{\includegraphics[width=0.9\textwidth]{est_p_one}}
\caption{Inferred polygenicity $p$ in simulations with continuous outcomes. Horizontal dashed lines represent the true simulated values. For LDpred2-auto, suffix ``nojump''/``jump'' refers to using \texttt{allow\_jump\_sign = FALSE/TRUE} (and \texttt{use\_MLE = TRUE}), and 
``noMLE'' refers to using \texttt{use\_MLE = FALSE} (and \texttt{allow\_jump\_sign = FALSE}). Note that the recommended option is to use \texttt{allow\_jump\_sign = FALSE} \cite[]{prive2021identifying}. \label{fig:simu_p}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=0.8\textwidth]{est_alpha_one}}
\caption{Inferred $\alpha$ in simulations with continuous outcomes. Horizontal dashed lines represent the true simulated values. Horizontal dotted lines represent boundaries imposed on the LDpred2-auto estimates. For LDpred2-auto, suffix ``nojump''/``jump'' refers to using \texttt{allow\_jump\_sign = FALSE/TRUE} (and \texttt{use\_MLE = TRUE}). Note that ``LDpred2\_noMLE'' (\texttt{use\_MLE = FALSE}) does not infer $\alpha$. Note that the recommended option is to use \texttt{allow\_jump\_sign = FALSE} \cite[]{prive2021identifying}. \label{fig:simu_alpha}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{postp_calib}}
\caption{Calibration of per-variant posterior probabilities of being causal in simulations with continuous outcomes. Green dashed lines represent the 1:1 line. Only results for $\alpha = -0.5$ and \texttt{allow\_jump\_sign = FALSE} are represented. \label{fig:simu_postp_calib}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{local_h2_calib}}
\caption{Calibration of per-block heritability estimates in simulations with continuous outcomes. Green dashed lines represent the 1:1 line. Only results for $\alpha = -0.5$ and \texttt{allow\_jump\_sign = FALSE} are represented. \label{fig:simu_h2_calib}}
\end{figure}

\begin{figure}[p]
	\centering
	\begin{subfigure}[b]{0.7\textwidth}
		\includegraphics[width=\textwidth]{susie_002}
		\caption{With $p=0.002$.}
		\vspace*{1em}
	\end{subfigure}
	\begin{subfigure}[b]{0.8\textwidth}
		\includegraphics[width=\textwidth]{susie_02}
		\vspace*{-2.6em}
		\caption{With $p=0.02$.}
		\vspace*{1em}		
	\end{subfigure}
	\caption{Calibration of per-variant posterior probabilities of being causal in two simulations with continuous outcomes, assuming $h^2=0.1$, $\alpha=-0.5$, and $N=10^5$.\label{fig:susie}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=0.9\textwidth]{est_r2_one}}
\caption{Inferred predictive performance $r^2$ in simulations with continuous outcomes. Green segments represent $r^2$ in the test set. For LDpred2-auto, suffix ``nojump''/``jump'' refers to using \texttt{allow\_jump\_sign = FALSE/TRUE} (and \texttt{use\_MLE = TRUE}), and 
``noMLE'' refers to using \texttt{use\_MLE = FALSE} (and \texttt{allow\_jump\_sign = FALSE}). Note that the recommended option is to use \texttt{allow\_jump\_sign = FALSE} \cite[]{prive2021identifying}. \label{fig:simu_r2}}
\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

%[ ADD THE SAME, BUT AVERAGING OVER 10 SIMUS -> CI OF THE MEAN ]

\FloatBarrier

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[p]
	\centerline{\includegraphics[width=0.82\textwidth]{est_binary_h2_one}}
	\caption{Inferred SNP heritability $h^2$ in simulations with binary outcomes. Horizontal dashed lines represent the true simulated values. The 95\% confidence interval for the LDpred2-auto estimate is obtained from the 2.5\% and 97.5\% quantiles of all the $h^2$ estimates from the iterations (after burn-in) of the chains kept. The 95\% confidence interval for the LD Score regression estimate is obtained from $\pm$1.96 times its standard error. {\NEFF} All $h^2$ estimates are transformed to the liability scale with K\_pop=K and K\_GWAS either K (the simulated prevalence) or 0.5 (when using Neff) using function \texttt{coef\_to\_liab} from R package bigsnpr \cite{lee2012better,prive2017efficient}. \label{fig:simu_h2_bin}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.85\textwidth]{est_binary_p_one}}
	\caption{Inferred polygenicity $p$ in simulations with binary outcomes. Horizontal dashed lines represent the true simulated values. {\NEFF} \label{fig:simu_p_bin}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.82\textwidth]{est_binary_alpha_one}}
	\caption{Inferred $\alpha$ in simulations with binary outcomes. Horizontal dashed lines represent the true simulated values. Horizontal dotted lines represent boundaries imposed on the estimates. {\NEFF} \label{fig:simu_alpha_bin}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.85\textwidth]{est_binary_r2_one}}
	\caption{Inferred predictive performance $r^2$ in simulations with binary outcomes. Green segments represent $r^2$ in the test set. {\NEFF} All $r^2$ estimates are transformed to the liability scale with K\_pop=K and K\_GWAS either K (the simulated prevalence) or 0.5 (when using Neff) using function \texttt{coef\_to\_liab} from R package bigsnpr \cite{lee2012better,prive2017efficient}; the ones from the test set are transformed with K and K. \label{fig:simu_r2_bin}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{ukbb_h2}}
	\caption{SNP heritability estimates from either LDpred2-auto or LD Score regression for all 248 phenotypes defined from the UK Biobank. These are stratified by the polygenicity estimated from LDpred2-auto. Green dashed lines represent the 1:1 line. The 95\% confidence interval for the LDpred2-auto estimate is obtained from the 2.5\% and 97.5\% quantiles of all the $h^2$ estimates from the iterations (after burn-in) of the chains kept. The 95\% confidence interval for the LD Score regression estimate is obtained from $\pm$1.96 times its standard error. \label{fig:ukbb_h2}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{ukbb_h2_p}}
	\caption{Estimates from LDpred2-auto for either the SNP heritability $h^2$ or the polygenicity $p$ for all 248 phenotypes defined from the UK Biobank. Estimates of $h^2$ are constrained to be at least 0.001, and at least $10^{-5}$ for $p$. The 95\% confidence interval for the LDpred2-auto estimate is obtained from the 2.5\% and 97.5\% quantiles of all the estimates from the iterations (after burn-in) of the chains kept. \label{fig:ukbb_h2_p}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{ukbb_h2_alpha}}
	\caption{Estimates from LDpred2-auto for either the SNP heritability $h^2$ or $\alpha$ for all 248 phenotypes defined from the UK Biobank. The 95\% confidence interval for the LDpred2-auto estimate is obtained from the 2.5\% and 97.5\% quantiles of all the estimates from the iterations (after burn-in) of the chains kept. We only show phenotypes for which there are more than 25 chains kept, because simulations have shown that $\alpha$ estimates are unreliable when a small number of chains is kept (Figure \ref{fig:simu_alpha}). \label{fig:ukbb_h2_alpha}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{ukbb_compare_hm3_small}}
\caption{LDpred2-auto estimates for UKBB phenotypes with either a small or a large LD reference. Only 154 phenotypes with more than 25 chains kept when using the large LD reference are represented here. Red dashed lines represent the 1:1 line. The 95\% confidence interval for the LDpred2-auto estimate (in green) is obtained from the 2.5\% and 97.5\% quantiles of all the estimates from the iterations (after burn-in) of the chains kept. The 95\% confidence interval for $r^2$ in the test set is obtained from bootstrap. \label{fig:small_LD}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{ukbb_compare_hm3_altpop}}
\caption{LDpred2-auto estimates for UKBB phenotypes with either a N.W. European or a S. European (``altpop'') LD reference. Only 154 phenotypes with more than 25 chains kept when using the N.W. European LD reference are represented here. Red dashed lines represent the 1:1 line. The 95\% confidence interval for the LDpred2-auto estimate (in green) is obtained from the 2.5\% and 97.5\% quantiles of all the estimates from the iterations (after burn-in) of the chains kept. The 95\% confidence interval for $r^2$ in the test set is obtained from bootstrap. \label{fig:alt_LD}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=0.85\textwidth]{ukbb_compare_mle}}
\caption{$r^2$ estimates (in the test set) for UKBB phenotypes, using LDpred2-auto with either the new sampling (MLE, 3-parameter model) or the previous one (2-parameter model). Red dashed lines represent the 1:1 line. The 95\% confidence interval for $r^2$ in the test set is obtained from bootstrap. \label{fig:compare_MLE}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.85\textwidth]{ukbb_compare_jumpsign}}
	\caption{LDpred2-auto estimates for 248 UKBB phenotypes, with or without option `allow\_jump\_sign' enabled. Red dashed lines represent the 1:1 line. The 95\% confidence interval for the LDpred2-auto estimate (in green) is obtained from the 2.5\% and 97.5\% quantiles of all the estimates from the iterations (after burn-in) of the chains kept. The 95\% confidence interval for $r^2$ in the test set is obtained from bootstrap. \label{fig:jumpsign}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.85\textwidth]{ukbb_compare_notransfo}}
	\caption{LDpred2-auto estimates for 49 log-transformed UKBB phenotypes, versus for their raw versions. Red dashed lines represent the 1:1 line. The 95\% confidence interval for the LDpred2-auto estimate (in green) is obtained from the 2.5\% and 97.5\% quantiles of all the estimates from the iterations (after burn-in) of the chains kept. The 95\% confidence interval for $r^2$ in the test set is obtained from bootstrap. \label{fig:notransfo}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.85\textwidth]{ukbb_compare_transfo}}
	\caption{LDpred2-auto estimates for 49 log-transformed UKBB phenotypes, versus for their rank-based inverse normal (RIN) transformed versions. Red dashed lines represent the 1:1 line. The 95\% confidence interval for the LDpred2-auto estimate (in green) is obtained from the 2.5\% and 97.5\% quantiles of all the estimates from the iterations (after burn-in) of the chains kept. The 95\% confidence interval for $r^2$ in the test set is obtained from bootstrap. \label{fig:RINT}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{ukbb_chains}}
	\caption{Distribution of the number of LDpred2-auto chains kept across 248 UKBB phenotypes. \label{fig:ukbb_chains}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=0.95\textwidth]{ukbb_runtimes}}
	\caption{Distribution of LDpred2-auto runtimes across 248 UKBB phenotypes. For each phenotype, 50 chains were used, parallelized over 13 cores. \label{fig:ukbb_runtimes}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{ukbb_local_h2}}
	\caption{Genome-wide and per-block heritability estimates from LDpred2-auto for UKBB phenotypes. The HapMap3+ variants are used here. The maximum local $h^2$ is the maximum heritability estimate across all 431 independent LD blocks defined for this set of variants. Only phenotypes for which this represents at least 10\% of the total heritability are represented. \label{fig:ukbb_local_h2}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{protein_local_h2}}
	\caption{Genome-wide and per-block heritability estimates from LDpred2-auto for 90 protein concentrations \cite[]{folkersen2020genomic}. The HapMap3+ variants are used here. The maximum local $h^2$ is the maximum heritability estimate across all 431 independent LD blocks defined for this set of variants. \label{fig:protein_local_h2}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{median_local_h2}}
	\caption{Per-block median heritability across 169 UKBB phenotypes. The HapMap3+ set of variants is used, with 431 independent LD blocks. Only phenotypes with more than 25 chains kept are used here. The top block is on chromosome 6 [22.1-41.4 Mb], which contains the HLA region. \label{fig:median_local_h2}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{median_postp}}
\caption{Per-variant median probabilities of being causal across 169 UKBB phenotypes. Variants with a median probability larger than 0.01 were mapped to genes using R package rsnps (only the largest per peak). The HapMap3+ set of variants is used. Only phenotypes with more than 25 chains kept are used here. \label{median_postp}}
\end{figure}


\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{postp_height}}
	\caption{Per-variant probabilities of being causal for height. These are provided by LDpred2-auto when using the external GWAS summary statistics of 1.6M European individuals from \citet{yengo2022saturated}. \label{fig:postp_height}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{height_enrichment}}
\caption{Heritability enrichment from LDpred2-auto for height across 50 functional annotations. The 95\% confidence interval for the LDpred2-auto estimate is obtained from the 2.5\% and 97.5\% quantiles of all the estimates from the iterations (after burn-in) of the chains kept. \label{fig:enrichment}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centering
	\begin{subfigure}[b]{0.85\textwidth}
		\includegraphics[width=\textwidth]{ldpred2_Asthma}
		\caption{Asthma}
		\vspace*{3em}
	\end{subfigure}
	\begin{subfigure}[b]{0.85\textwidth}
		\includegraphics[width=\textwidth]{ldpred2_BrCa}
		\caption{Breast cancer}
		\vspace*{3em}		
	\end{subfigure}
\end{figure}
\begin{figure}[p]\ContinuedFloat
	\centering
	\begin{subfigure}[b]{0.85\textwidth}
		\includegraphics[width=\textwidth]{ldpred2_CAD}
		\caption{Coronary artery disease}
		\vspace*{3em}		
	\end{subfigure}
	\begin{subfigure}[b]{0.85\textwidth}
		\includegraphics[width=\textwidth]{ldpred2_MDD}
		\caption{Depression}
		\vspace*{3em}		
	\end{subfigure}
\end{figure}
\begin{figure}[p]\ContinuedFloat
	\centering
	\begin{subfigure}[b]{0.85\textwidth}
		\includegraphics[width=\textwidth]{ldpred2_PrCa}
		\caption{Prostate cancer}
		\vspace*{3em}		
	\end{subfigure}
	\begin{subfigure}[b]{0.85\textwidth}
		\includegraphics[width=\textwidth]{ldpred2_T1D}
		\caption{Type 1 diabetes}
		\vspace*{3em}		
	\end{subfigure}
\end{figure}
\begin{figure}[p]\ContinuedFloat
	\centering
	\begin{subfigure}[b]{0.8\textwidth}
	\includegraphics[width=\textwidth]{ldpred2_T2D}
	\caption{Type 2 diabetes}
	\vspace*{3em}		
	\end{subfigure}
	\begin{subfigure}[b]{0.8\textwidth}
	\includegraphics[width=\textwidth]{ldpred2_VitaminD}
	\caption{Vitamin D}
	\vspace*{1em}		
	\end{subfigure}
	\caption{LDpred2-auto results with external GWAS summary statistics. The UK Biobank is used as test set to compute $r^2$.\label{fig:sumstats}}
\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier
\clearpage

\bibliographystyle{natbib}
\bibliography{refs}

\end{document}
