%% LyX 1.3 created this file.  For more info, see http://www.lyx.org/.
%% Do not edit unless you really know what you are doing.
\documentclass[english, 12pt]{article}
\usepackage{times}
%\usepackage{algorithm2e}
\usepackage{url}
\usepackage{bbm}
\usepackage[T1]{fontenc}
\usepackage[latin1]{inputenc}
\usepackage{geometry}
\geometry{verbose,letterpaper,tmargin=2cm,bmargin=2cm,lmargin=1.5cm,rmargin=1.5cm}
\usepackage{rotating}
\usepackage{color}
\usepackage{graphicx}
\usepackage{amsmath, amsthm, amssymb}
\usepackage{setspace}
\usepackage{lineno}
\usepackage{hyperref}
\usepackage{bbm}
\usepackage{makecell}
\usepackage{placeins}
\usepackage{subcaption}

%\renewcommand{\arraystretch}{1.8}

%\linenumbers
%\doublespacing
\onehalfspacing
\usepackage[numbers,super]{natbib}

%Pour les rajouts
\usepackage{color}
\definecolor{trustcolor}{rgb}{0,0,1}

\usepackage{dsfont}
\usepackage[warn]{textcomp}
\usepackage{adjustbox}
\usepackage{multirow}
\usepackage{subcaption}
\usepackage{graphicx}
\graphicspath{{../figures/}}
\DeclareMathOperator*{\argmin}{\arg\!\min}

\let\tabbeg\tabular
\let\tabend\endtabular
\renewenvironment{tabular}{\begin{adjustbox}{max width=0.95\textwidth}\tabbeg}{\tabend\end{adjustbox}}

\makeatletter

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% LyX specific LaTeX commands.
%% Bold symbol macro for standard LaTeX users
%\newcommand{\boldsymbol}[1]{\mbox{\boldmath $#1$}}

%% Because html converters don't know tabularnewline
\providecommand{\tabularnewline}{\\}
\renewcommand*{\arraystretch}{1.2}

\usepackage{babel}
\makeatother


\begin{document}

\renewcommand{\thefigure}{S\arabic{figure}}
\setcounter{figure}{0}
\renewcommand{\thetable}{S\arabic{table}}
\setcounter{table}{0}
\renewcommand{\theequation}{S\arabic{equation}}
\setcounter{equation}{0}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{est_h2_one}}
\caption{Inferred SNP heritability $h^2$ in simulations with continuous outcomes. Horizontal lines represent the true simulated values. The 95\% confidence interval for the LDpred2-auto estimate is obtained from the 2.5\% and 97.5\% quantiles of all the $h^2$ estimates from the iterations (after burn-in) of the chains kept. The 95\% confidence interval for the LD Score regression estimate is obtained from $\pm$1.96 times its standard error. \label{fig:simu_h2}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{est_p_one}}
\caption{Inferred polygenicity $p$ in simulations with continuous outcomes. \label{fig:simu_p}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{est_alpha_one}}
\caption{Inferred $\alpha$ in simulations with continuous outcomes. \label{fig:simu_alpha}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{postp_calib}}
\caption{Calibration of per-variant posterior probabilities of being causal in simulations with continuous outcomes. \label{fig:simu_postp_calib}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{local_h2_calib}}
\caption{Calibration of per-block heritability estimates in simulations with continuous outcomes. \label{fig:simu_h2_calib}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{est_r2_one}}
\caption{Inferred predictive performance $r^2$ in simulations with continuous outcomes. \label{fig:simu_r2}}
\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

%[ ADD THE SAME, BUT AVERAGING OVER 10 SIMUS -> CI OF THE MEAN ]

\FloatBarrier

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{est_binary_h2_one}}
	\caption{All h2 estimates were transformed to the liability scale with K\_pop=K and K\_GWAS either K or 0.5 (when using Neff) \label{}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{est_binary_p_one}}
	\caption{ \label{}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{est_binary_alpha_one}}
	\caption{ \label{}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{est_binary_r2_one}}
	\caption{Transformed both r2 to the liability scale, where the inferred one using Neff used K\_GWAS=0.5  \label{}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{binary_postp_calib}}
	\caption{ \label{}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=0.9\textwidth]{ukbb_h2}}
	\caption{SNP heritability estimates from either LDpred2-auto or LD Score regression for all 248 phenotypes defined from the UK Biobank. These are stratified by estimated polygenicity (from LDpred2-auto). The 95\% confidence interval for the LDpred2-auto estimate is obtained from the 2.5\% and 97.5\% quantiles of all the $h^2$ estimates from the iterations (after burn-in) of the chains kept. The 95\% confidence interval for the LD Score regression estimate is obtained from $\pm$1.96 times its standard error. \label{fig:ukbb_h2}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{ukbb_h2_p}}
	\caption{Estimates from LDpred2-auto for either the SNP heritability $h^2$ or the polygenicity $p$ for all 248 phenotypes defined from the UK Biobank. The 95\% confidence intervals for the LDpred2-auto estimates are obtained from the 2.5\% and 97.5\% quantiles of all the estimates from the iterations (after burn-in) of the chains kept. \label{fig:ukbb_h2_p}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{ukbb_h2_alpha}}
	\caption{Estimates from LDpred2-auto for either the SNP heritability $h^2$ or $\alpha$ for all 248 phenotypes defined from the UK Biobank. The 95\% confidence intervals for the LDpred2-auto estimates are obtained from the 2.5\% and 97.5\% quantiles of all the estimates from the iterations (after burn-in) of the chains kept. We only show phenotypes for which there are more than 25 chains kept, because simulations have shown that $\alpha$ estimates are unreliable when a small number of chains is kept (Figure \ref{fig:simu_alpha}). \label{fig:ukbb_h2_alpha}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{ukbb_r2}}
	\caption{Inferred predictive performance $r^2$ from the Gibbs sampler of LDpred2-auto versus the ones obtained in the test set, for all 248 phenotypes defined from the UK Biobank. These are stratified by estimated heritability (from LDpred2-auto). The 95\% confidence interval for the LDpred2-auto estimate is obtained from the 2.5\% and 97.5\% quantiles of all the $r^2$ estimates from the iterations (after burn-in) of the chains kept. The 95\% confidence interval for $r^2$ in the test set is obtained from bootstrap. \label{fig:ukbb_r2}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{ukbb_compare_hm3_small}}
\caption{ \label{fig:small_LD}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{ukbb_compare_hm3_altpop}}
\caption{ \label{fig:alt_LD}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{ukbb_compare_hm3_plus}}
	\caption{ \label{fig:hm3_plus}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{ukbb_chains}}
\caption{Distribution of the number of LDpred2-auto chains kept across 248 UKBB phenotypes. \label{fig:ukbb_chains}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{ukbb_runtimes}}
\caption{Distribution of LDpred2-auto runtimes across 248 UKBB phenotypes. For each phenotype, 50 chains were used, parallelized over 13 cores. \label{fig:ukbb_runtimes}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{ukbb_compare_notransfo}}
	\caption{ \label{fig:notransfo}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{ukbb_compare_transfo}}
	\caption{ \label{fig:RINT}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{ukbb_local_h2}}
	\caption{ \label{fig:ukbb_local_h2}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{protein_local_h2}}
	\caption{ \label{fig:protein_local_h2}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{median_local_h2}}
	\caption{Per-block median heritability across 169 UKBB phenotypes. The HapMap3+ set of variants is used, with 431 independent LD blocks. Only phenotypes with more than 25 chains kept are used here. The top block is on chromosome 6 [22.1-41.4 Mb], which contains the HLA region. \label{fig:median_local_h2}}
\end{figure}

\begin{figure}[p]
\centerline{\includegraphics[width=\textwidth]{median_postp}}
\caption{Per-variant median probabilities of being causal across 169 UKBB phenotypes. The HapMap3+ set of variants is used. Only phenotypes with more than 25 chains kept are used here. \label{median_postp}}
\end{figure}

\begin{figure}[p]
	\centerline{\includegraphics[width=\textwidth]{height_enrichment}}
	\caption{Heritability enrichment from LDpred2-auto for height across 50 functional annotations. $N$ is the GWAS sample size. \label{fig:enrichment}}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\FloatBarrier
\clearpage

\bibliographystyle{natbib}
\bibliography{refs}

\end{document}
